### PR TITLE
Fix undefined

### PR DESCRIPTION
**Pull request by sthaeron**
Fixes various edge cases relating to undefined functions for processes that do not use type declarations. The pr also makes various simplifications and improvements to the compiler.

- Simplifies process application and system output pattern matching.
- Refactors how type-level lambdas are ignored based on work done by @klarasm.
- Adds pattern matching to support systems that use undefined functions with no type declarations.
- Adds SDF examples to test changes.
- Removes unnecessary modules from LSP executable in cabal file.

Note that there might be additional edge cases relating to undefined functions without type declarations. Nevertheless the changes from this pr works for all current SDF examples and the newly added ones.

### CODE REVIEW

File: src/CoreIRToForSyDeIR.hs [hunk 1]
```diff
@@ -97,6 +96,7 @@ translateSystem initialContext expr = case expr of
     let context1 = translateSystemBinds initialContext [(b, e)]
         context2 = translateSystem context1 out
      in context2
+  -- _ -> error ("\nSystem Inputs: " ++ show (systemInputs initialContext) ++ "\nSystem Outputs: " ++ show (systemOutputs initialContext))
```
Comment (klarasm) [inline src/CoreIRToForSyDeIR.hs]: Should likely be removed in the final version

File: forsyde-devtools.cabal [hunk 2]
```diff
@@ -53,6 +53,11 @@ library
     SDFSchedule
     Utilities
 
+library examples
+  import: common-options
+  hs-source-dirs: examples
+  visibility: private
+
```
Comment (klarasm) [inline forsyde-devtools.cabal:59]: This is really nice, thanks!

File: forsyde-devtools.cabal [hunk 3]
```diff
@@ -1,4 +1,4 @@
-cabal-version: 2.4
+cabal-version: 3.0
```
Comment (klarasm) [inline forsyde-devtools.cabal:1]: Do we need it for the other part below or is this an unrelated change?

File: src/CoreIRToForSyDeIR.hs [hunk 4]
```diff
@@ -230,84 +248,31 @@ translateSystemBinds initialcontext = foldl' translateSystemBind initialcontext
...
-     in (binder, context1)
   Case (Var i) _ _ alts ->
     let bindingId = IRVar i
         index = getIndexFromAlts initialContext alts
         binder = Binding bindingId index
      in (binder, initialContext)
-  _ -> error ("translateSystemExpr - Unsupported CoreExpr:\n" ++ prettyCoreExpr (flags initialContext) expr)
+  _ -> case translateConstructorApplication expr of
+    Just (pcId, argumentIds) ->
+      -- error ("pcId: " ++ show (pcId) ++ " argumentIds: " ++ show (argumentIds))
```
Comment (klarasm) [inline src/CoreIRToForSyDeIR.hs]: Should probably be removed in the final version.

File: src/CoreIRToForSyDeIR.hs [hunk 5]
```diff
@@ -230,84 +248,31 @@ translateSystemBinds initialcontext = foldl' translateSystemBind initialcontext
...
+translateConstructorApplication :: CoreExpr -> Maybe (IRId, [IRId])
+translateConstructorApplication expr = aux expr []
+  where
+    aux currentExpr arguments = case currentExpr of
+      Lam _ e -> aux e arguments
+      App e (Var a) | isId a -> aux e ((IRVar a) : arguments)
+      App e (Type _) -> aux e arguments
+      App e (App (Var a) (Type _)) -> aux e ((IRVar a) : arguments)
+      Var i -> Just (IRVar i, arguments)
+      _ -> Nothing
```
Comment (klarasm) [inline src/CoreIRToForSyDeIR.hs]: This is really nice!

File: src/CoreIRToForSyDeIR.hs [hunk 6]
```diff
@@ -134,7 +153,7 @@ translateSystemOutputs initialContext expr = case expr of
      in context3
   -- System has 4 outputs
   App (App (App (App (App (App (App (App (Var _) (Type _)) (Type _)) (Type _)) (Type _)) (Var out1)) (Var out2)) (Var out3)) (Var out4) ->
-    let context1 = foldl updateSystemOutput initialContext [out1, out2, out3, out4]
+    let context1 = foldl updateSystemOutput initialContext [(IRVar out1), (IRVar out2), (IRVar out3), (IRVar out4)]
```
Comment (klarasm) [inline src/CoreIRToForSyDeIR.hs]: We might as well change these to `foldl'` when we are at it.

File: src/CoreIRToForSyDeIR.hs [hunk 7]
```diff
@@ -106,15 +106,34 @@ translateSystem initialContext expr = case expr of
...
+        context3 = context2 {systemInputs = reverse (systemInputs context2)}
+        context4 = updateConstructorsAndSignals context3
+     in context4
+  Case (Let (Rec binds) out) _ _ _alts ->
+    let context1 = translateSystemBinds initialContext binds
+        context2 = translateSystemOutputs context1 out
+     in -- outId = getOutputFromAlts initialContext alts
+        -- context2 = updateSystemOutput context1 outId
+        -- context3 = context2 {systemInputs = reverse (systemInputs context2)}
+        -- context4 = updateConstructorsAndSignals context3
```
Comment (klarasm) [inline src/CoreIRToForSyDeIR.hs]: Should probably be removed in the final version.

File: src/CoreIRToForSyDeIR.hs [hunk 8]
```diff
@@ -106,15 +106,34 @@ translateSystem initialContext expr = case expr of
 -- translation to support netlists with additional system ouputs.
 translateSystemOutputs :: TranslationContext -> CoreExpr -> TranslationContext
 translateSystemOutputs initialContext expr = case expr of
+  Let (NonRec b e) _out ->
```
Comment (klarasm) [inline src/CoreIRToForSyDeIR.hs]: Might be nice with a comment here (and below) to mirror the other matches.

**Review by klarasm — APPROVED** [review]
Looks good!
